lib/AST/ASTMangler.cpp

```diff
@@ -2167,9 +2167,29 @@ ASTMangler::getSpecialManglingContext(const ValueDecl *decl,
   if (isa<TypeDecl>(decl)) {
     if (auto *clangDecl = cast_or_null<clang::NamedDecl>(decl->getClangDecl())){
       bool hasNameForLinkage;
-      if (auto *tagDecl = dyn_cast<clang::TagDecl>(clangDecl))
+      if (auto *tagDecl = dyn_cast<clang::TagDecl>(clangDecl)) {
         hasNameForLinkage = tagDecl->hasNameForLinkage();
```

Comment:
I haven't made the change in this patch (yet anyways), but I think we should stop making naked calls to `tagDecl->hasNameForLinkage` and `tagDecl->getTypedefNameForAnonDecl`. This is the second patch I've made fixing problems because these methods don't behave as we expect. Basically every time the fix is to add some version of this extra check I've added here below. I think we should consider writing wrappers around these two methods and using them everywhere in the ClangImporter.

Along a similar vein, I think once we feel we've rounded up all the CF_OPTION related problems and the CF_Availability.h header change is stable, we should consider a refactor to consolidate these hacks. Ideally,
I would like some sort of Clang AST transformation so that once the ClangImporter gets its hands on it, the Objective-C and C++ definitions of CF_OPTIONS are the same. I imagine this is not the last time we will get divergences in the SDK, a more robust mechanism might be in order.